synapse/storage/databases/main/events.py

```diff
@@ -1199,22 +1220,31 @@ def _filter_events_and_contexts_for_duplicates(
     def _update_room_depths_txn(
         self,
         txn,
+        *,
         events_and_contexts: List[Tuple[EventBase, EventContext]],
-        backfilled: bool,
+        use_negative_stream_ordering: bool = False,
     ):
         """Update min_depth for each room
 
         Args:
             txn (twisted.enterprise.adbapi.Connection): db connection
             events_and_contexts (list[(EventBase, EventContext)]): events
                 we are persisting
-            backfilled (bool): True if the events were backfilled
+            use_negative_stream_ordering: Whether to start stream_ordering on
+                the negative side and decrement. This should be set as True
+                for backfilled events because backfilled events get a negative
+                stream ordering so they don't come down incremental `/sync`.
         """
         depth_updates: Dict[str, int] = {}
         for event, context in events_and_contexts:
             # Remove the any existing cache entries for the event_ids
             txn.call_after(self.store._invalidate_get_event_cache, event.event_id)
-            if not backfilled:
+            # This will update the `stream_ordering` position to mark the latest
+            # event as the front of the room. This should not be done for
+            # backfilled events because backfilled events have negative
+            # stream_ordering and happened in the past so we know that we don't
+            # need to update the stream_ordering tip for the room.
+            if not use_negative_stream_ordering:
```

Comment:
can we just infer this from `event.internal_metadata.stream_ordering >= 0` ?

Reply:
This seems reasonable 👍 Cool simplification we're seeing across the arguments here!